File: admin/src/api/gameRank.ts
import apiClient from './client';
import type { ApiResponse, Pagination } from '../types/api';

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Game Category - 游戏分类模型
 * 对应后端 model.GameCategory
 */
export interface GameCategory {
    id: number;
    name: string;
    description?: string;
    iconUrl?: string;
    sortOrder: number;
    isActive: boolean;
    createdAt: string;
    updatedAt: string;
    // 关联数据（非持久化字段）
    games?: Game[];
    serviceItems?: any[];
}

/**
 * Game Rank - 游戏段位配置模型
 * 对应后端 model.GameRank
 */
export interface GameRank {
    id: number;
    gameId: number;
    name: string;
    level: number;
    priceCents: number;
    iconUrl?: string;
    color?: string;
    description?: string;
    sortOrder: number;
    isActive: boolean;
    createdAt: string;
    updatedAt: string;
    // 关联数据（非持久化字段）
    game?: Game;
    gameName?: string;
}

/**
 * Game - 游戏基础信息（段位关联时返回）
 */
export interface Game {
    id: number;
    key: string;
    name: string;
    category?: string;
    categoryId?: number;
    iconUrl?: string;
    coverUrl?: string;
    description?: string;
    isActive: boolean;
    sortOrder: number;
}

// ============================================================================
// Request/Response DTOs - Game Category
// ============================================================================

/**
 * 创建游戏分类请求
 */
export interface CreateGameCategoryDto {
    name: string;
    description?: string;
    iconUrl?: string;
    sortOrder?: number;
}

/**
 * 更新游戏分类请求
 */
export interface UpdateGameCategoryDto {
    name?: string;
    description?: string;
    iconUrl?: string;
    sortOrder?: number;
    isActive?: boolean;
}

/**
 * 批量更新分类状态请求
 */
export interface BatchUpdateCategoryStatusDto {
    categoryIds: number[];
    isActive: boolean;
}

/**
 * 批量删除分类请求
 */
export interface BatchDeleteCategoriesDto {
    categoryIds: number[];
}

/**
 * 游戏分类查询参数
 */
export interface GameCategoryQueryParams {
    page?: number;
    pageSize?: number;
    keyword?: string;
    isActive?: boolean;
}

/**
 * 批量操作响应
 */
export interface BatchOperationResponse {
    successCount: number;
    failureCount: number;
    errors?: string[];
}

// ============================================================================
// Request/Response DTOs - Game Rank
// ============================================================================

/**
 * 创建游戏段位请求
 */
export interface CreateGameRankDto {
    gameId: number;
    name: string;
    level?: number;
    priceCents?: number;
    iconUrl?: string;
    color?: string;
    description?: string;
    sortOrder?: number;
    isActive?: boolean;
}

/**
 * 更新游戏段位请求
 */
export interface UpdateGameRankDto {
    name?: string;
    level?: number;
    priceCents?: number;
    iconUrl?: string;
    color?: string;
    description?: string;
    sortOrder?: number;
    isActive?: boolean;
}

/**
 * 批量删除段位请求
 */
export interface BatchDeleteGameRanksDto {
    ids: string[];
}

/**
 * 批量更新段位状态请求
 */
export interface BatchUpdateRankStatusDto {
    ids: string[];
    isActive: boolean;
}

/**
 * 游戏段位查询参数
 */
export interface GameRankQueryParams {
    page?: number;
    pageSize?: number;
    gameId?: number;
    keyword?: string;
    isActive?: boolean;
}

// ============================================================================
// API Client - Game Category
// ============================================================================

/**
 * 获取游戏分类列表
 * GET /admin/game-categories
 */
const getGameCategories = (
    params?: GameCategoryQueryParams
): Promise<ApiResponse<GameCategory[]>> => {
    return apiClient.get('/admin/game-categories', { params }).then(res => res.data);
};

/**
 * 获取单个游戏分类
 * GET /admin/game-categories/{id}
 */
const getGameCategory = (
    id: number
): Promise<ApiResponse<GameCategory>> => {
    return apiClient.get(`/admin/game-categories/${id}`).then(res => res.data);
};

/**
 * 创建游戏分类
 * POST /admin/game-categories
 */
const createGameCategory = (
    data: CreateGameCategoryDto
): Promise<ApiResponse<GameCategory>> => {
    return apiClient.post('/admin/game-categories', data).then(res => res.data);
};

/**
 * 更新游戏分类
 * PUT /admin/game-categories/{id}
 */
const updateGameCategory = (
    id: number,
    data: UpdateGameCategoryDto
): Promise<ApiResponse<GameCategory>> => {
    return apiClient.put(`/admin/game-categories/${id}`, data).then(res => res.data);
};

/**
 * 删除游戏分类
 * DELETE /admin/game-categories/{id}
 */
const deleteGameCategory = (
    id: number
): Promise<ApiResponse<void>> => {
    return apiClient.delete(`/admin/game-categories/${id}`).then(res => res.data);
};

/**
 * 批量更新游戏分类状态
 * POST /admin/game-categories/batch/status
 */
const batchUpdateCategoryStatus = (
    data: BatchUpdateCategoryStatusDto
): Promise<ApiResponse<BatchOperationResponse>> => {
    return apiClient.post('/admin/game-categories/batch/status', data).then(res => res.data);
};

/**
 * 批量删除游戏分类
 * POST /admin/game-categories/batch/delete
 */
const batchDeleteCategories = (
    data: BatchDeleteCategoriesDto
): Promise<ApiResponse<BatchOperationResponse>> => {
    return apiClient.post('/admin/game-categories/batch/delete', data).then(res => res.data);
};

// ============================================================================
// API Client - Game Rank
// ============================================================================

/**
 * 获取游戏段位列表
 * GET /admin/game-ranks
 */
const getGameRanks = (
    params?: GameRankQueryParams
): Promise<ApiResponse<GameRank[]>> => {
    return apiClient.get('/admin/game-ranks', { params }).then(res => res.data);
};

/**
 * 根据游戏ID获取段位列表
 * GET /admin/games/{gameId}/ranks
 */
const getGameRanksByGame = (
    gameId: number
): Promise<ApiResponse<GameRank[]>> => {
    return apiClient.get(`/admin/games/${gameId}/ranks`).then(res => res.data);
};

/**
 * 获取单个游戏段位
 * GET /admin/game-ranks/{id}
 */
const getGameRank = (
    id: number
): Promise<ApiResponse<GameRank>> => {
    return apiClient.get(`/admin/game-ranks/${id}`).then(res => res.data);
};

/**
 * 创建游戏段位
 * POST /admin/game-ranks
 */
const createGameRank = (
    data: CreateGameRankDto
): Promise<ApiResponse<GameRank>> => {
    return apiClient.post('/admin/game-ranks', data).then(res => res.data);
};

/**
 * 更新游戏段位
 * PUT /admin/game-ranks/{id}
 */
const updateGameRank = (
    id: number,
    data: UpdateGameRankDto
): Promise<ApiResponse<GameRank>> => {
    return apiClient.put(`/admin/game-ranks/${id}`, data).then(res => res.data);
};

/**
 * 删除游戏段位
 * DELETE /admin/game-ranks/{id}
 */
const deleteGameRank = (
    id: number
): Promise<ApiResponse<void>> => {
    return apiClient.delete(`/admin/game-ranks/${id}`).then(res => res.data);
};

/**
 * 批量删除游戏段位
 * POST /admin/game-ranks/batch/delete
 */
const batchDeleteGameRanks = (
    data: BatchDeleteGameRanksDto
): Promise<ApiResponse<{ deleted: number }>> => {
    return apiClient.post('/admin/game-ranks/batch/delete', data).then(res => res.data);
};

/**
 * 批量更新游戏段位状态
 * POST /admin/game-ranks/batch/status
 */
const batchUpdateRankStatus = (
    data: BatchUpdateRankStatusDto
): Promise<ApiResponse<{ updated: number }>> => {
    return apiClient.post('/admin/game-ranks/batch/status', data).then(res => res.data);
};

// ============================================================================
// Combined API Export
// ============================================================================

/**
 * Game Rank & Category API
 * 统一导出所有游戏分类和段位相关接口
 */
export const gameRankApi = {
    // Game Category APIs
    getGameCategories,
    getGameCategory,
    createGameCategory,
    updateGameCategory,
    deleteGameCategory,
    batchUpdateCategoryStatus,
    batchDeleteCategories,

    // Game Rank APIs
    getGameRanks,
    getGameRanksByGame,
    getGameRank,
    createGameRank,
    updateGameRank,
    deleteGameRank,
    batchDeleteGameRanks,
    batchUpdateRankStatus,
};

// Re-export types for convenience
export type {
    GameCategory,
    GameRank,
    Game,
    CreateGameCategoryDto,
    UpdateGameCategoryDto,
    BatchUpdateCategoryStatusDto,
    BatchDeleteCategoriesDto,
    GameCategoryQueryParams,
    BatchOperationResponse,
    CreateGameRankDto,
    UpdateGameRankDto,
    BatchDeleteGameRanksDto,
    BatchUpdateRankStatusDto,
    GameRankQueryParams,
};

export default gameRankApi;

File: api/internal/handler/admin/gameRank.go
package admin

import (
	"strconv"

	"github.com/gin-gonic/gin"

	"gamelink/internal/handler/middleware"
	"gamelink/internal/model"
	"gamelink/internal/repository"
	"gamelink/internal/service/gamerank"
	"gamelink/pkg/apierr"
)

// GameRankHandler 游戏段位配置管理接口
type GameRankHandler struct {
	svc *gamerank.GameRankService
}

// NewGameRankHandler 创建Handler
func NewGameRankHandler(svc *gamerank.GameRankService) *GameRankHandler {
	return &GameRankHandler{svc: svc}
}

// GameRankPayload 创建/更新段位请求体
type GameRankPayload struct {
	GameID      uint64 `json:"gameId" binding:"required"`
	Name        string `json:"name" binding:"required,max=64"`
	Level       int    `json:"level"`
	PriceCents  int64  `json:"priceCents"`
	IconURL     string `json:"iconUrl"`
	Color       string `json:"color"`
	Description string `json:"description"`
	SortOrder   int    `json:"sortOrder"`
	IsActive    bool   `json:"isActive"`
}

// GameRankUpdatePayload 更新段位请求体（不含GameID）
type GameRankUpdatePayload struct {
	Name        string `json:"name" binding:"required,max=64"`
	Level       int    `json:"level"`
	PriceCents  int64  `json:"priceCents"`
	IconURL     string `json:"iconUrl"`
	Color       string `json:"color"`
	Description string `json:"description"`
	SortOrder   int    `json:"sortOrder"`
	IsActive    bool   `json:"isActive"`
}

// ListGameRanks
// @Summary      列出游戏段位
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Param        page       query  int     false  "页码"
// @Param        pageSize   query  int     false  "每页数量"
// @Param        gameId     query  int     false  "游戏ID筛选"
// @Param        keyword    query  string  false  "关键词搜索"
// @Param        isActive   query  bool    false  "是否启用"
// @Produce      json
// @Success      200  {object}  model.APIResponse[[]model.GameRank]
// @Router       /admin/game-ranks [get]
func (h *GameRankHandler) ListGameRanks(c *gin.Context) {
	page, pageSize, ok := parsePagination(c)
	if !ok {
		return
	}

	gameID, ok := QueryUint64PtrAndRespond(c, "gameId", apierr.ErrInvalidGameID)
	if !ok {
		return
	}

	var isActive *bool
	if v := c.Query("isActive"); v != "" {
		b := v == "true"
		isActive = &b
	}

	opts := repository.GameRankListOptions{
		Page:     page,
		PageSize: pageSize,
		GameID:   gameID,
		Keyword:  c.Query("keyword"),
		IsActive: isActive,
	}

	ranks, pagination, err := h.svc.ListPaged(c.Request.Context(), opts)
	if err != nil {
		respondError(c, err)
		return
	}
	respondList(c, ranks, pagination)
}

// ListGameRanksByGame
// @Summary      根据游戏ID获取段位列表
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Param        gameId   path  int  true  "游戏ID"
// @Produce      json
// @Success      200  {object}  model.APIResponse[[]model.GameRank]
// @Router       /admin/games/{gameId}/ranks [get]
func (h *GameRankHandler) ListGameRanksByGame(c *gin.Context) {
	gameID, ok := ParseIDAndRespond(c, "gameId")
	if !ok {
		return
	}

	ranks, err := h.svc.ListByGameID(c.Request.Context(), gameID)
	if err != nil {
		respondError(c, err)
		return
	}
	respondSuccess(c, ranks)
}

// GetGameRank
// @Summary      获取游戏段位详情
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Param        id   path  int  true  "段位ID"
// @Produce      json
// @Success      200  {object}  model.APIResponse[model.GameRank]
// @Failure      404  {object}  model.ErrorResponse
// @Router       /admin/game-ranks/{id} [get]
func (h *GameRankHandler) GetGameRank(c *gin.Context) {
	id, ok := ParseIDAndRespond(c, "id")
	if !ok {
		return
	}

	rank, err := h.svc.Get(c.Request.Context(), id)
	if err != nil {
		respondError(c, err)
		return
	}
	respondSuccess(c, rank)
}

// CreateGameRank
// @Summary      创建游戏段位
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Accept       json
// @Produce      json
// @Param        request  body  GameRankPayload  true  "段位信息"
// @Success      201  {object}  model.APIResponse[model.GameRank]
// @Failure      400  {object}  model.ErrorResponse
// @Router       /admin/game-ranks [post]
func (h *GameRankHandler) CreateGameRank(c *gin.Context) {
	var payload GameRankPayload
	if !ValidateAndRespond(c, &payload) {
		return
	}

	rank, err := h.svc.Create(c.Request.Context(), gamerank.CreateInput{
		GameID:      payload.GameID,
		Name:        payload.Name,
		Level:       payload.Level,
		PriceCents:  payload.PriceCents,
		IconURL:     payload.IconURL,
		Color:       payload.Color,
		Description: payload.Description,
		SortOrder:   payload.SortOrder,
		IsActive:    payload.IsActive,
	})
	if err != nil {
		respondError(c, err)
		return
	}
	respondCreated(c, rank)
}

// UpdateGameRank
// @Summary      更新游戏段位
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Accept       json
// @Produce      json
// @Param        id       path  int                    true  "段位ID"
// @Param        request  body  GameRankUpdatePayload  true  "段位信息"
// @Success      200  {object}  model.APIResponse[model.GameRank]
// @Failure      404  {object}  model.ErrorResponse
// @Router       /admin/game-ranks/{id} [put]
func (h *GameRankHandler) UpdateGameRank(c *gin.Context) {
	id, ok := ParseIDAndRespond(c, "id")
	if !ok {
		return
	}

	var payload GameRankUpdatePayload
	if !ValidateAndRespond(c, &payload) {
		return
	}

	rank, err := h.svc.Update(c.Request.Context(), id, gamerank.UpdateInput{
		Name:        payload.Name,
		Level:       payload.Level,
		PriceCents:  payload.PriceCents,
		IconURL:     payload.IconURL,
		Color:       payload.Color,
		Description: payload.Description,
		SortOrder:   payload.SortOrder,
		IsActive:    payload.IsActive,
	})
	if err != nil {
		respondError(c, err)
		return
	}
	respondUpdated(c, rank)
}

// DeleteGameRank
// @Summary      删除游戏段位
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Param        id   path  int  true  "段位ID"
// @Produce      json
// @Success      200  {object}  model.SuccessResponse
// @Failure      404  {object}  model.ErrorResponse
// @Router       /admin/game-ranks/{id} [delete]
func (h *GameRankHandler) DeleteGameRank(c *gin.Context) {
	id, ok := ParseIDAndRespond(c, "id")
	if !ok {
		return
	}

	if err := h.svc.Delete(c.Request.Context(), id); err != nil {
		respondError(c, err)
		return
	}
	respondDeleted(c)
}

// BatchDeleteGameRanksRequest 批量删除请求
type BatchDeleteGameRanksRequest struct {
	IDs []string `json:"ids" binding:"required,min=1"`
}

// BatchDeleteGameRanks
// @Summary      批量删除游戏段位
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Accept       json
// @Produce      json
// @Param        request  body  BatchDeleteGameRanksRequest  true  "段位ID列表"
// @Success      200  {object}  model.APIResponse[map[string]int64]
// @Failure      400  {object}  model.ErrorResponse
// @Router       /admin/game-ranks/batch/delete [post]
func (h *GameRankHandler) BatchDeleteGameRanks(c *gin.Context) {
	var req BatchDeleteGameRanksRequest
	if !ValidateAndRespond(c, &req) {
		return
	}

	ids := make([]uint64, 0, len(req.IDs))
	for _, idStr := range req.IDs {
		id, err := strconv.ParseUint(idStr, 10, 64)
		if err != nil {
			respondBadRequest(c, "invalid id: "+idStr)
			return
		}
		ids = append(ids, id)
	}

	deleted, err := h.svc.BatchDelete(c.Request.Context(), ids)
	if err != nil {
		respondError(c, err)
		return
	}

	respondSuccess(c, map[string]int64{"deleted": deleted})
}

// BatchUpdateStatusRequest 批量更新状态请求
type BatchUpdateStatusRequest struct {
	IDs      []string `json:"ids" binding:"required,min=1"`
	IsActive bool     `json:"isActive"`
}

// BatchUpdateGameRankStatus
// @Summary      批量更新游戏段位状态
// @Tags         Admin/GameRanks
// @Security     BearerAuth
// @Accept       json
// @Produce      json
// @Param        request  body  BatchUpdateStatusRequest  true  "请求体"
// @Success      200  {object}  model.APIResponse[map[string]int64]
// @Failure      400  {object}  model.ErrorResponse
// @Router       /admin/game-ranks/batch/status [post]
func (h *GameRankHandler) BatchUpdateGameRankStatus(c *gin.Context) {
	var req BatchUpdateStatusRequest
	if !ValidateAndRespond(c, &req) {
		return
	}

	ids := make([]uint64, 0, len(req.IDs))
	for _, idStr := range req.IDs {
		id, err := strconv.ParseUint(idStr, 10, 64)
		if err != nil {
			respondBadRequest(c, "invalid id: "+idStr)
			return
		}
		ids = append(ids, id)
	}

	updated, err := h.svc.BatchUpdateStatus(c.Request.Context(), ids, req.IsActive)
	if err != nil {
		respondError(c, err)
		return
	}

	respondSuccess(c, map[string]int64{"updated": updated})
}

// RegisterGameRankRoutes 注册游戏段位管理路由
func RegisterGameRankRoutes(router gin.IRouter, svc *gamerank.GameRankService, pm *middleware.PermissionMiddleware) {
	h := NewGameRankHandler(svc)

	group := router.Group("/game-ranks")
	group.Use(pm.RequireAuth())
	{
		group.GET("", pm.RequirePermission(model.HTTPMethodGET, "/api/v1/admin/game-ranks"), h.ListGameRanks)
		group.POST("", pm.RequirePermission(model.HTTPMethodPOST, "/api/v1/admin/game-ranks"), h.CreateGameRank)
		group.GET("/:id", pm.RequirePermission(model.HTTPMethodGET, "/api/v1/admin/game-ranks/:id"), h.GetGameRank)
		group.PUT("/:id", pm.RequirePermission(model.HTTPMethodPUT, "/api/v1/admin/game-ranks/:id"), h.UpdateGameRank)
		group.DELETE("/:id", pm.RequirePermission(model.HTTPMethodDELETE, "/api/v1/admin/game-ranks/:id"), h.DeleteGameRank)
		group.GET("/game/:gameId", pm.RequirePermission(model.HTTPMethodGET, "/api/v1/admin/game-ranks/game/:gameId"), h.ListGameRanksByGame)
		group.POST("/batch/delete", pm.RequirePermission(model.HTTPMethodPOST, "/api/v1/admin/game-ranks/batch/delete"), h.BatchDeleteGameRanks)
		group.POST("/batch/status", pm.RequirePermission(model.HTTPMethodPOST, "/api/v1/admin/game-ranks/batch/status"), h.BatchUpdateGameRankStatus)
	}
}
